### ISSUE TITLE
Acknowledging a comment shouldn't take me out of Blocking

### ISSUE THREAD

**Issue by kballard** [Reviewable/Reviewable#392]
The description for "Blocking" says I'll stay in that state until I manually change my disposition. But Acknowledging a reply in the discussion immediately moves me back to Discussing (current status: OK) and resolves the discussion! This is very surprising. I should be able to acknowledge a comment without taking it out of Blocking.

**Comment by kballard** [comment]
For reference, I want to acknowledge it to get rid of the "New" badge.

**Comment by pkaminski** [comment]
That would indeed be quite bad, but I can't reproduce it -- can you provide more details, including a pointer to the discussion?

Looking at the code for Acknowledge, it will only change a user's disposition if it's not already active (one of `discussing`, `blocking`, or `satisfied`), and even then it will only change it to `following`, never `discussing`.  So at this point I'm stumped as to what could've happened.

**Comment by kballard** [comment]
Weird. The discussion in question is https://reviewable.io/reviews/postmates/getitnow-ios/1138. It's got a ton of comments on it, and it actually has another comment that's bugged but I didn't file a ticket about it yet because it hasn't reproduced beyond that one comment (namely, it's a comment that shows up in my "to reply" list even though I made the last comment in the discussion).

At this point, the comment that caused me to file the ticket has been successfully resolved so it's not blocking. But I do have another blocking comment at the moment.

**Comment by pkaminski** [comment]
That's a reference to the review, not a specific discussion...  Also, it's possible for a discussion to be unreplied even if you have the last comment -- this happens when you're a reviewer in `discussing` state and the author has acknowledged your comment, so the only thing preventing the discussion from resolving is your not-OK `discussing` disposition.  If that doesn't describe your case please point me to the discussion so I can investigate!

**Comment by kballard** [comment]
Oh I see. Unfortunately there's so many comments on that review that I'm not sure how to find the one I had previously been marked as blocking on.

The discussion that I'm currently marked as blocking on is https://reviewable.io/reviews/postmates/getitnow-ios/1138#-KPoYHSRUjd3KYhaVLN2. I can't be positive that the issue happened with that one, but I feel like it might have. I know I saw the issue twice, I just can't remember if it was twice on the same comment, or once on each comment.

The "bugged" discussion is https://reviewable.io/reviews/postmates/getitnow-ios/1138#-KQSNaEvwDSKS7wmVuTw. I am in the "discussing" state and the PR author has acknowledged my comment. However, the PR author hasn't responded, so they're just in the "following" state, and there's no action I can take on my end (short of manually changing my disposition to Satisfied).

**Comment by pkaminski** [comment]
It's very likely that https://reviewable.io/reviews/postmates/getitnow-ios/1138#-KPoYHSRUjd3KYhaVLN2 was an instance of the situation I described above, working as expected.

Unfortunately  https://reviewable.io/reviews/postmates/getitnow-ios/1138#-KQSNaEvwDSKS7wmVuTw is no longer in the state you described -- it took me a while to get back to this issue, sorry!  But you're correct that an Acknowledge will not join the discussion, and as long as the discussion has only one active participant then the only way to close it is to switch your disposition to Satisfied (or Withdrawn, if available).

So basically, everything is working as expected AFAICT.  But a more interesting question is: how would you *like* it to work?  Specifically, it seems that the behavior when the reviewer opens a discussion and the author just Acknowledges without replying is not intuitive for you.  Any suggestions on how to improve it?  Thanks.

**Comment by kballard** [comment]
The problem with the "bugged" discussion isn't that it wasn't resolved, it's that it continued to show up as a discussion that I needed to take action on, despite there being nothing for me to acknowledge or respond to. Having the discussion stay open when acknowledged is fine, it just shouldn't show up in the red count on my end when that happens. After all, if the author simply acknowledged it, it means they haven't done anything besides read the comment, so it's really up to them to actually respond.

**Comment by pkaminski** [comment]
I understand your argument since that used to be my logic as well... until another user pointed out that this resulted in an unresolved discussion that wasn't red for *anyone*.  My underlying philosophy has always been that everything that's blocking completion should always be red for at least one person so I added an exceptional condition in this case to put the red on the reviewer.  The only other option I can think of would be to prohibit the author from acknowledging a reviewer-initiated discussion that they're not yet active in, and that has no other participants, but this felt very heavy-handed ("it stays red until you reply!").

With the current approach, the red count tells you that you need to either unilaterally resolve the discussion, or prompt the author for further action.

**Comment by kballard** [comment]
Acknowledge serves double-duty right now. It both adjusts disposition if you're in "discussing", potentially resolving the issue, and it also gets rid of the New badge on a comment. I think this is a bit of a problem; there's been plenty of times where I wanted to rid of the New badge, but acknowledging it would resolve the discussion, and the discussion isn't ready to resolve (e.g. the PR author replies with something like "You're right, I'll fix that" and leaves themselves in discussing (Currently: OK), so the underlying issue isn't actually fixed yet but if I acknowledge the discussion gets resolved). In those situations I end up having to mark myself as Blocking so I can acknowledge it to get rid of the New badge, but it's really easy to accidentally resolve it first.

All this is to say, if I make a comment and the PR author acknowledges, but does not reply, that most likely means they were just trying to get rid of the "New" badge. In this case, I think it's perfectly reasonable to leave the red on them instead of on the comment author; the PR author hasn't made any changes or responded, so it's pretty weird to demand the comment author do something. If the PR author fixes the issue, they should say "Done", and if they don't want to make a code change, they should reply to that effect.

I would also like a way for a comment author to remove the new badge on a PR author's response without changing their disposition (and I guess leaving the comment red for them), but I'm not sure how to do that without complicating the UI and making it potentially confusing, so this is really just a "if you can think of a way to do this that's not potentially confusing" suggestion.

**Comment by pkaminski** [comment]
Agreed, the semantics you propose (author keeps the red marker) make more sense.  Fixed, hopefully -- these conditions are getting pretty darn tricky.

I couldn't think of any way to give you control of the New marker separately from acknowledging without further complicating an already complex UI.  The current data schema would support such a differentiation, though, so if an idea presents itself it should be pretty easy to add.
